docs: package addons (#3892)
aws/copilot-cli: site/content/docs/developing/package-addons.en.md

> # Package Addons <span class="version" > added in [v1.21.0](../../blogs/release-v121.en.md) </span>

Reviewer: I understand if it's deemed too big and we want a separate page for this, in that scenario should we create a new section under developing like this:
```
    - Developing:
      - Additional AWS Resources:
          - Modeling With AWS CloudFormation: docs/developing/addons/modeling.en.md
          - Packaging Local Artifacts: docs/developing/addons/package.en.md
```
We would need to add a redirect as well so that the old URL reroutes to the new hierarchy:
```
    - redirects:
        redirect_maps:
            'docs/developing/additional-aws-resources.en.md': 'docs/developing/addons/modeling.en.md'
```

Copilot supports uploading local files referenced from your addon templates to S3, and replacing the relevant resource properties with the uploaded S3 location.

> On [`copilot svc deploy`](../commands/svc-deploy.en.md) or [`copilot svc package --upload-assets`](../commands/svc-package.en.md), certain fields on supported resources will be updated with an S3 locaton.

Reviewer 2: ```suggestion
On [`copilot svc deploy`](../commands/svc-deploy.en.md) or [`copilot svc package --upload-assets`](../commands/svc-package.en.md), certain fields on supported resources will be updated with an S3 location.
```

Author: :sob:

To see the full list of resources that are supported, take a look at the [AWS CLI documentation](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/cloudformation/package.html).

This feature can be used to deploy local Lambda Functions stored in the same repo as another Copilot service.
For example, to deploy a javascript Lambda Function alongside a copilot service, you can add this resource to your [addon template](./additional-aws-resources.en.md):

???+ note "Example Lambda Function"
    === "copilot/service-name/addons/lambda.yml"

        ```yaml hl_lines="4"
          ExampleFunction:
            Type: AWS::Lambda::Function
            Properties:
              Code: lambdas/example/
              Handler: "index.handler"
              Timeout: 900
              MemorySize: 512
              Role: !GetAtt "ExampleFunctionRole.Arn"
              Runtime: nodejs16.x

>           ExampleFunctionRole:

Reviewer: nit: should we remove the role here, just to highlight more the relevant properties `Handler` and `Code`?

Author: probably a good idea in this section - i'll leave them in the full example below, though.

        ```js
        exports.handler = function (event, context) {
	        console.log('example event:', event);
	        context.succeed('success!');
        };
        ```

On `copilot svc deploy`, the `lambdas/example` directory will be zipped and uploaded to S3, and the `Code` property will be updated to:
```yaml
Code:
  S3Bucket: copilotBucket
  S3Key: hashOfLambdasExampleZip
```
before the addon template is uploaded and deployed by Copilot.
If you specify a file, the file is directly uploaded to S3.
If you specify a folder, the folder will be zipped before being uploaded to S3.
For some resources that require a zip (e.g., `AWS::Serverless::Function`), a file will be zipped before upload as well.

File paths are considered relative to the parent of the `copilot/` directory in your repo.
For the above example, the folder structure would look like:
```bash
.
├── copilot
│   └── example-service
│       ├── addons
│       │   └── lambda.yml
│       └── manifest.yml
└── lambdas
    └── example
        └── index.js
```
Absolute paths are supported as well, though they may not work as well across multiple machines.

## Example: DynamoDB Stream Processing Lambda

> This example will walk through creating a [Dynamo DB](https://aws.amazon.com/dynamodb/) table with a lambda function conntected to process events from the [table's stream](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Streams.html).

Reviewer 2: ```suggestion
This example will walk through creating a [Dynamo DB](https://aws.amazon.com/dynamodb/) table with a lambda function connected to process events from the [table's stream](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Streams.html).
```

Author: ⚰️

> This example will walk through creating a [Dynamo DB](https://aws.amazon.com/dynamodb/) table with a lambda function conntected to process events from the [table's stream](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Streams.html).

Reviewer: ```suggestion
This example will walk through creating an [Amazon DynamoDB](https://aws.amazon.com/dynamodb/) table with a lambda function conntected to process events from the [table's stream](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Streams.html).
```

> This architecture could be useful if you have a service that needs to minimize latency on storing data, but can kick off separate process that takes longer to process the data.

Reviewer 2: ???
```suggestion
This architecture could be useful if you have a service that needs to minimize latency on storing data, but can kick off a separate process that takes longer to process the data.
```

#### Prerequesites:
- [A deployed copilot service](../concepts/services.en.md)

#### Steps:
1. Generate a DynamoDB table addon for your service by running `copilot storage init` (More info [here!](./storage.en.md))
2. Add the [`StreamSpecification`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-dynamodb-table.html#cfn-dynamodb-table-streamspecification) property to the generated [`AWS::DynamoDB::Table`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-dynamodb-table.html) resource:

>   ```yaml title="copilot/service-name/addons/ddb.yml"

Reviewer: :TIL: `title` and `hl_lines` 😍
